cvmfs/object_fetcher.h

```diff
@@ -0,0 +1,387 @@
+               object_hash.ToString().c_str(), dest.c_str());
+      return false;
+    }
+
+    if (! zlib::DecompressPath2Path(source, dest)) {
+      LogCvmfs(kLogDownload, kLogDebug, "failed to extract object %s from '%s' "
+                                        "to '%s' (errno: %d)",
+               object_hash.ToString().c_str(), source.c_str(), dest.c_str(),
+               errno);
+      return false;
+    }
+
+    *file_path = dest;
+    return true;
+  }
+
+
+ protected:
+  std::string BuildPath(const std::string &relative_path) const {
+    return base_path_ + "/" + relative_path;
+  }
+
+  std::string BuildPath(const shash::Any    &hash,
+                        const shash::Suffix  suffix) const {
+    return BuildPath("data" + hash.MakePathWithSuffix(1, 2, suffix));
+  }
+
+ private:
+  const std::string base_path_;
+  const std::string temporary_directory_;
+};
+
+template <class CatalogT, class HistoryT>
+struct object_fetcher_traits<LocalObjectFetcher<CatalogT, HistoryT> > {
+    typedef CatalogT catalog_t;
+    typedef HistoryT history_t;
+};
+
+
+/**
+ * This implements the AbstractObjectFetcher<> to retrieve repository objects
+ * from a remote location through HTTP. It verifies the repository's signature
+ * and the downloaded data integrity.
+ */
+template <class CatalogT = catalog::Catalog,
+          class HistoryT = history::SqliteHistory>
+class HttpObjectFetcher :
+  public AbstractObjectFetcher<HttpObjectFetcher<CatalogT, HistoryT> >
+{
+ protected:
+  typedef HttpObjectFetcher<CatalogT, HistoryT>  this_t;
+  typedef AbstractObjectFetcher<this_t>          base_t;
+
+ public:
+  /**
+   * Factory method to instatiate an HttpObjectFetcher<>. This implementation
+   * contains internal state that cannot be copied and needs destruction. Hence
+   * it can only reside on the heap and the user is reponsible to delete the
+   * instance after usage.
+   *
+   * @param repo_name  the name of the repository to download objects from
+   * @param repo_url   the URL to the repository's backend storage
+   * @param repo_keys  a list of public keys to access the repository
+   * @param temp_dir   location to store decompressed tmp data
+   * @return           a HttpObjectFetcher<> object or NULL on error
+   */
+  static this_t* Create(const std::string &repo_name,
+                        const std::string &repo_url,
+                        const std::string &repo_keys,
+                        const std::string &temp_dir) {
+    UniquePtr<this_t> fetcher(new this_t(repo_name,
+                                         repo_url,
+                                         repo_keys,
+                                         temp_dir));
+    assert (fetcher);
+    return (fetcher->Initialize()) ? fetcher.Release() : NULL;
+  }
+
+  ~HttpObjectFetcher() {
+    signature_manager_.Fini();
+    download_manager_.Fini();
+  }
+
+ protected:
+  bool Initialize() {
+    download_manager_.Init(1, true);
```

Comment:
That is a good point, in fact. It adds some [boiler plate](https://github.com/reneme/cvmfs/commit/7732ec1c5c4069db2659490f19a9a9441f2a165f) to the user side, since one needs to initialise these objects every time an `HttpObjectFetcher<>` is needed. However, the gained flexibility is a real advantage. [Fixed](https://github.com/reneme/cvmfs/commit/f7dea8145991de0f4093f8a9911c6fce3002ade0).